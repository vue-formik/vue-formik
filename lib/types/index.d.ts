import { ObjectSchema } from "yup";

interface FormikHelpers<T> {
  reset: () => void;
  setErrors: (errors: Partial<Record<keyof T, string | Record<string, string>>>) => void;
  setValues: (values: T) => void;
  setSubmitting: (value: boolean) => void;
}

type ValidationRule = (
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  value: any,
) =>
  | string
  | undefined
  | (string | undefined)[]
  | Record<string, string | undefined>
  | Record<string, string | undefined>[];

type FormikValidationSchema<T extends object> =
  | ObjectSchema<T>
  | Partial<Record<keyof T, ValidationRule>>;

type FormikOnSubmit<T> = (values: T, helpers: FormikHelpers<T>) => void;
