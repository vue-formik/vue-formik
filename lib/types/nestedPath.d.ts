/**
 * Type to generate all possible nested paths in an object, including arrays.
 */
// type NestedPaths<T> = T extends object
//   ? {
//     [K in keyof T]: K extends string
//       ? T[K] extends (infer U)[]
//         ?
//         | `${K}`
//         | `${K}[${number}]`
//         | `${K}[${number}]${NestedArrayPaths<U>}`
//         | `${K}.${NestedPaths<T[K]>}`
//         : `${K}` | `${K}.${NestedPaths<T[K]>}`
//       : never;
//   }[keyof T]
//   : never;
//
// type NestedArrayPaths<U> = U extends (infer V)[]
//   ? `[${number}]` | `[${number}]${NestedArrayPaths<V>}`
//   : U extends object
//     ? `[${number}].${NestedPaths<U>}`
//     : never;
//
// /**
//  * Type to extract the nested value type given a path.
//  */
// type NestedValue<T, P extends string> = P extends keyof T
//   ? T[P]
//   : P extends `${infer K}[${infer I}]${infer Rest}`
//     ? K extends keyof T
//       ? T[K] extends (infer U)[]
//         ? I extends `${number}`
//           ? NestedValue<U, Rest extends `.${infer R}` ? R : Rest>
//           : undefined
//         : undefined
//       : K extends ""
//         ? T extends (infer U)[]
//           ? I extends `${number}`
//             ? NestedValue<U, Rest extends `.${infer R}` ? R : Rest>
//             : undefined
//           : undefined
//         : undefined
//     : P extends `${infer K}.${infer Rest}`
//       ? K extends keyof T
//         ? NestedValue<T[K], Rest>
//         : undefined
//       : undefined;

type NestedPaths<T> = T extends object
  ? {
    [K in keyof T]: K extends string
      ? T[K] extends (infer U)[]
        ?
        | `${K}`
        | `${K}[${number}]`
        | `${K}[${number}]${NestedArrayPaths<U>}`
        | `${K}.${NestedPaths<T[K]>}`
        : `${K}` | `${K}.${NestedPaths<T[K]>}`
      : never;
  }[keyof T]
  : never;

type NestedArrayPaths<U> = U extends (infer V)[]
  ? `[${number}]` | `[${number}]${NestedArrayPaths<V>}`
  : U extends object
    ? `[${number}].${NestedPaths<U>}`
    : never;

type NestedValue<T, P extends string> = P extends keyof T
  ? T[P]
  : P extends `${infer K}[${infer I}]${infer Rest}`
    ? K extends keyof T
      ? T[K] extends (infer U)[] | undefined
        ? I extends `${number}`
          ? NestedValue<Exclude<U, undefined>, Rest extends `.${infer R}` ? R : Rest>
          : never
        : never
      : K extends ""
        ? T extends (infer U)[] | undefined
          ? I extends `${number}`
            ? NestedValue<Exclude<U, undefined>, Rest extends `.${infer R}` ? R : Rest>
            : never
          : never
        : never
    : P extends `${infer K}.${infer Rest}`
      ? K extends keyof T
        ? NestedValue<T[K], Rest>
        : T extends Record<string, any>
          ? any
          : never
      : any;
