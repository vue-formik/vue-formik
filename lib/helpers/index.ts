import updateNestedProperty from "@/helpers/updateNestedProperty";
import getNestedValue from "@/helpers/getNestedValue";
import clearReactiveObject from "@/helpers/clearReactiveObject";
import { constructLabel } from "@/helpers/utils";

export {
  updateNestedProperty,
  getNestedValue,
  clearReactiveObject,
  constructLabel,
};
