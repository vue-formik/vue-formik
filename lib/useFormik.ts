import { computed, reactive, toRaw, watch } from 'vue'
import { ObjectSchema } from 'yup'
import type { FormikHelpers, IUseFormik, ValidationRule } from '@/types'
import { getNestedValue, updateNestedProperty } from '@/helpers'

const useFormik = <T extends object>({
  initialValues,
  validationSchema,
  onSubmit,
}: IUseFormik<T>) => {
  const validate = (values: T): Partial<Record<keyof T, string>> => {
    const validationErrors: Partial<Record<keyof T, string>> = {};

    if (validationSchema instanceof ObjectSchema) {
      try {
        const vSchema = validationSchema as ObjectSchema<T>;
        vSchema.validateSync(values, { abortEarly: false });
      } catch (e) {
        const err = e as never as { inner: { path: string; message: string }[] };
        if (typeof err === "object" && Array.isArray(err?.inner)) {
          err.inner.forEach((error) => {
            validationErrors[error.path as keyof T] = error.message;
          });
        }
      }
    } else {
      for (const key in validationSchema) {
        const value = values[key as keyof T];
        const rule = validationSchema[key as keyof T] as ValidationRule<T[keyof T]>;

        if (typeof rule === 'function') {
          const error = rule(value);
          if (error) {
            validationErrors[key as keyof T] = error;
          }
        }
      }
    }

    return validationErrors;
  };

  const values = reactive(initialValues);
  const errors = reactive(validate(initialValues));
  const touched = reactive({} as Partial<Record<keyof T, boolean | Record<string, boolean>>>);

  const setValues = (newValues: T) => {
    Object.assign(values, newValues); // Safely copy values into reactive object
    Object.assign(errors, validate(newValues)); // Validate updated values
  };

  const setErrors = (newErrors: T) => {
    Object.assign(errors, newErrors);
  };

  const reset = () => {
    Object.assign(values, initialValues);
    Object.assign(errors, validate(initialValues));
  };

  const setFieldValue = (field: string, value: unknown) => {
    updateNestedProperty(values as Record<string, unknown>, field, value);
  };

  const setFieldTouched = (field: string, touchedValue: boolean) => {
    updateNestedProperty(touched as Record<string, unknown>, field, touchedValue);
  }

  const handleSubmit = (e: Event) => {
    e.preventDefault();
    onSubmit(
      toRaw(values) as T,
      {
        setErrors,
        reset,
        setValues,
      } as FormikHelpers<T>
    );
  };

  const handleBlur = (e: FocusEvent) => {
    const fieldName = (e.target as HTMLInputElement).name as keyof T;
    setFieldTouched(fieldName.toString(), true);
  }
  const handleChange = (e: Event) => {
    const target = e.target as HTMLInputElement;
    const fieldName = target.name as keyof T;
    const value = target.type === 'checkbox' ? target.checked : target.value;

    setFieldValue(fieldName.toString(), value);
  }

  const isValid = computed(() => {
    return Object.keys(errors).length === 0;
  })

  const isDirty = computed(() => {
    return JSON.stringify(values) !== JSON.stringify(initialValues);
  })

  watch(
    () => values,
    () => {
      const validationErrors = validate(toRaw(values) as T);
      //@ts-expect-error Object.keys(errors) is not iterable
      Object.keys(errors).forEach((key) => delete errors[key]); // Clear existing errors
      Object.assign(errors, validationErrors); // Assign new errors
    },
    { deep: true }
  );


  const hasFieldError = (field: string) => {
    const errorValue = getNestedValue(errors, field);
    const touchedValue = getNestedValue(touched, field);
    return (errorValue !== undefined && !!touchedValue);
  }

  const getFieldError = (field: string) => {
    if (hasFieldError(field)) {
      return errors[field as keyof typeof errors];
    } else {
      return '';
    }
  }

  return {
    values,
    errors,
    touched,
    setValues,
    reset,
    setFieldValue,
    setFieldTouched,
    handleBlur,
    handleChange,
    handleSubmit,
    hasFieldError,
    getFieldError,
    isValid,
    isDirty
  };
};

export default useFormik;
