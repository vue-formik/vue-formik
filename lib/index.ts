import useFormik from "@/composables/useFormik";
import FormTextArea from "@/components/FormTextArea.vue";
import FormTextField from "@/components/FormTextField.vue";

export { FormTextArea, FormTextField, useFormik };
