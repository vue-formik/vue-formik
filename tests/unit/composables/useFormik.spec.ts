import { test, expect, describe, vi } from "vitest";
import { useFormik } from "../../../lib";
import { nextTick } from "vue";
import { FormikHelpers } from "../../../lib/types";

describe("useFormik", () => {
  const initialValues = {
    name: "Kiran",
    email: "kiran@parajuli.cc",
    addresses: ["Home", "Office", ""],
  };
  const emptyInitialValues = {
    name: "",
    email: "",
    addresses: ["", "", ""],
  };
  const onSubmit = vi.fn((_values, { setSubmitting }) => {
    setSubmitting(false); // Simulate the completion of the submission process
  });
  const validationSchema = {
    name: (value: string) => (!value ? "Name is required" : undefined),
    email: (value: string) =>
      !value ? "Email is required" : !value.includes("@") ? "Invalid email" : undefined,
    addresses: (value: string[]) => {
      if (!Array.isArray(value)) {
        return "Addresses must be an array";
      }

      if (value.length === 0) {
        return "Address is required";
      }

      const errs = []
      for (let i = 0; i < value.length; i++) {
        if (!value[i]) {
          errs[i] = "Address is required";
        }
      }

      return errs.length ? errs : undefined;
    },
  };

  describe("submit", () => {
    test("should call onSubmit with initial values", () => {
      const { handleSubmit } = useFormik({ initialValues, onSubmit });
      handleSubmit({ preventDefault: vi.fn() });
      expect(onSubmit).toHaveBeenCalledWith(initialValues, expect.any(Object));
    });

    test("'isSubmitting' should return true when form is submitting", async () => {
      const onSubmit = vi.fn((_values, { setSubmitting }: FormikHelpers<typeof initialValues>) => {
        expect(isSubmitting.value).toBe(true);
        setSubmitting(false);
        expect(isSubmitting.value).toBe(false);
      });

      const { isSubmitting, handleSubmit } = useFormik({
        initialValues,
        onSubmit,
      });

      expect(isSubmitting.value).toBe(false);

      handleSubmit({ preventDefault: vi.fn() }); // Simulate form submission

      await nextTick();

      expect(isSubmitting.value).toBe(false); // Verify submission state is reset
    });
  });

  describe("validation", () => {
    test("should handle validation schema errors", async () => {
      const { errors, handleBlur, handleChange } = useFormik({
        initialValues: emptyInitialValues,
        onSubmit,
        validationSchema,
      });

      expect(errors).toEqual({
        name: "Name is required",
        email: "Email is required",
        addresses: [
          "Address is required",
          "Address is required",
          "Address is required",
        ]
      });

      handleChange({ target: { name: "name", value: "Kiran" } });
      handleBlur({ target: { name: "name" } });

      await nextTick();

      expect(errors.name).toBeUndefined();
      expect(errors.email).toBe("Email is required");
    });

    test("should handle non-array values for addresses field", () => {
      const { errors } = useFormik({
        initialValues: { addresses: "Invalid Value" }, // Pass a non-array value
        onSubmit,
        validationSchema,
      });

      expect(errors.addresses).toBe("Addresses must be an array");
    });

    test("should handle empty validation schema", () => {
      const { errors } = useFormik({ initialValues, onSubmit });
      expect(errors).toEqual({});
    });
  });

  describe("state updates", () => {
    test("should update touched fields", async () => {
      const { setFieldTouched, touched } = useFormik({ initialValues, onSubmit });
      setFieldTouched("name", true);

      await nextTick();

      expect(touched.name).toBe(true);
    });

    test("should update field values", () => {
      const { setFieldValue, values } = useFormik({ initialValues, onSubmit });
      setFieldValue("name", "Updated Name");

      expect(values.name).toBe("Updated Name");
    });

    test("should reset form state", async () => {
      const {
        reset,
        values,
        touched,
        setFieldValue,
        setFieldTouched
      } = useFormik({ initialValues, onSubmit });

      setFieldValue("name", "New Name");
      setFieldTouched("name", true);

      await nextTick();

      reset();

      await nextTick();

      expect(values).toEqual(initialValues);
      expect(touched).toEqual({});
    });
  });

  describe("field errors", () => {
    test("should return field error if present", async () => {
      const { getFieldError, setFieldValue, setFieldTouched } = useFormik({
        initialValues,
        onSubmit,
        validationSchema,
      });

      setFieldValue("name", "");
      setFieldTouched("name", true);

      await nextTick();

      expect(getFieldError("name")).toBe("Name is required");
    });

    test("should return true for hasFieldError if error exists", async () => {
      const { hasFieldError, setFieldValue, setFieldTouched } = useFormik({
        initialValues,
        onSubmit,
        validationSchema,
      });

      setFieldValue("name", "");
      setFieldTouched("name", true);

      await nextTick();

      expect(hasFieldError("name")).toBe(true);
    });
  });
});
