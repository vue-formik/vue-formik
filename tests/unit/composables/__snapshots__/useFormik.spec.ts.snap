// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useFormik > should initialize form with initial values 1`] = `
{
  "email": "kiran@parajuli.cc",
  "name": "Kiran",
}
`;

exports[`useFormik > should reset form 1`] = `
{
  "email": "updated@mail.cc",
  "name": "Updated Name",
}
`;

exports[`useFormik > should reset form 2`] = `
{
  "email": true,
  "name": true,
}
`;

exports[`useFormik > should reset form 3`] = `
{
  "email": "kiran@parajuli.cc",
  "name": "Kiran",
}
`;

exports[`useFormik > should reset form 4`] = `{}`;

exports[`useFormik > should set errors 1`] = `
{
  "name": "Value must be less than 255 characters",
}
`;

exports[`useFormik > should set field touched 1`] = `
{
  "name": true,
}
`;

exports[`useFormik > should set field value 1`] = `
{
  "email": "kiran@parajuli.cc",
  "name": "Updated Name",
}
`;

exports[`useFormik > should set values 1`] = `
{
  "email": "kiran@parajuli.cc",
  "name": "Updated Name",
}
`;

exports[`useFormik > should validate form with validation schema 1`] = `
{
  "email": "Email is required",
  "name": "Name is required",
}
`;

exports[`useFormik > should validate form with validation schema 2`] = `{}`;

exports[`useFormik > should validate form with validation schema 3`] = `
{
  "email": "Email is required",
}
`;

exports[`useFormik > should validate form with validation schema 4`] = `
{
  "name": true,
}
`;

exports[`useFormik > should validate form with validation schema 5`] = `
{
  "email": "Invalid email",
}
`;

exports[`useFormik > should validate form with validation schema 6`] = `
{
  "email": true,
  "name": true,
}
`;

exports[`useFormik > should validate form with validation schema 7`] = `{}`;

exports[`useFormik > should validate form with validation schema 8`] = `
{
  "email": true,
  "name": true,
}
`;
